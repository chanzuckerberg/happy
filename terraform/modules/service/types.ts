export type ECSComputeLimit = { cpu: ".25 vcpu", mem: "512" } |
{ cpu: ".25 vcpu", mem: "1 GB" } |
{ cpu: ".25 vcpu", mem: "2 GB" } |
{ cpu: ".5 vcpu", mem: "1 GB" } |
{ cpu: ".5 vcpu", mem: "2 GB" } |
{ cpu: ".5 vcpu", mem: "3 GB" } |
{ cpu: ".5 vcpu", mem: "4 GB" } |
{ cpu: "1 vcpu", mem: "2 GB" } |
{ cpu: "1 vcpu", mem: "3 GB" } |
{ cpu: "1 vcpu", mem: "4 GB" } |
{ cpu: "1 vcpu", mem: "5 GB" } |
{ cpu: "1 vcpu", mem: "6 GB" } |
{ cpu: "1 vcpu", mem: "7 GB" } |
{ cpu: "1 vcpu", mem: "8 GB" } |
{ cpu: "2 vcpu", mem: "4 GB" } |
{ cpu: "2 vcpu", mem: "5 GB" } |
{ cpu: "2 vcpu", mem: "6 GB" } |
{ cpu: "2 vcpu", mem: "7 GB" } |
{ cpu: "2 vcpu", mem: "8 GB" } |
{ cpu: "2 vcpu", mem: "9 GB" } |
{ cpu: "2 vcpu", mem: "10 GB" } |
{ cpu: "2 vcpu", mem: "11 GB" } |
{ cpu: "2 vcpu", mem: "12 GB" } |
{ cpu: "2 vcpu", mem: "13 GB" } |
{ cpu: "2 vcpu", mem: "14 GB" } |
{ cpu: "2 vcpu", mem: "15 GB" } |
{ cpu: "2 vcpu", mem: "16 GB" } |
{ cpu: "1 vcpu", mem: "4 GB" }

export enum ServiceType {
    PRIVATE = "private",
    INTERNAL = "internal",
    EXTERNAL = "external",
}

export enum Environment {
    RDEV = "rdev",
    DEV = "dev",
    STAGING = "staging",
    PROD = "prod",
}

export enum AWSRegion {
    EAST1 = "us-east-1",
    EAST2 = "us-east-2",
    WEST1 = "us-west-1",
    WEST2 = "us-west-2",
}

export interface ECSServiceDefinition {
    name: string
    // the number of replicas
    desiredCount: number
    // the port to expose
    port: number
    // the image to use in the container
    image: string
    // deployment size (only valid combinations allowed)
    computeLimits: ECSComputeLimit
    // if the service is on the internet, protected by Okta, or only exposed within the cluster
    serviceType: ServiceType
    healthCheckPath: string
    environment: EnvironmentVariables
}

export type EnvironmentVariables = { name: string, value: string }[]

export interface AWSLogConfig {
    logDriver: "awslogs",
    options: {
        "awslogs-stream-prefix": string,
        "awslogs-group": string,
        "awslogs-region": AWSRegion,
    }
}

export interface ContainerDefinition {
    name: string,
    image: string,
    memory: string,
    portMappings: {
        containerPort: number,
        hostPort?: number,
        protocol?: string
    }[]
    logConfiguration?: AWSLogConfig
    essential?: boolean,
    environment?: EnvironmentVariables
    command?: string
}

export interface HappyServiceMeta {
    env: Environment
    stackName: string
    region: AWSRegion
    serviceDef: ECSServiceDefinition
}
