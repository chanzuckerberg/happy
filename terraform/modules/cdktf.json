{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "514a6ca1-8aff-418a-86f4-7b8beea67509",
  "sendCrashReports": "true",
  "terraformProviders": ["aws@~> 4.0"],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
