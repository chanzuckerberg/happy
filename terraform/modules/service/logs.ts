import { Construct } from "constructs";
import { CloudwatchLogGroup } from "@cdktf/provider-aws/lib/cloudwatch-log-group";
import { AWSLogConfig, HappyServiceMeta } from "./types";

interface HappyLogsProps {
    meta: HappyServiceMeta,
    tags?: { [key: string]: string };
}

export class HappyLogs extends Construct {
    readonly logGroup: CloudwatchLogGroup
    readonly logConfig: AWSLogConfig
    constructor(scope: Construct, id: string, config: HappyLogsProps) {
        super(scope, id)

        this.logGroup = new CloudwatchLogGroup(scope, `cwg`, {
            retentionInDays: 365,
            name: `/happy/${config.meta.env}/${config.meta.stackName}/${config.meta.serviceDef.name}`,
            tags: config.tags,
        })

        this.logConfig = {
            logDriver: 'awslogs',
            options: {
                "awslogs-group": this.logGroup.id,
                "awslogs-region": config.meta.region,
                "awslogs-stream-prefix": config.meta.env,
            }
        }
    }
}
