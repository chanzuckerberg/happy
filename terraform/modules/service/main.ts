import { Construct } from "constructs";
import { App, TerraformOutput, TerraformStack, TerraformVariable } from "cdktf";
import { DataAwsRoute53Zone } from "@cdktf/provider-aws/lib/data-aws-route53-zone";
import { Route53Record } from "@cdktf/provider-aws/lib/route53-record";
import { DataAwsAlb } from "@cdktf/provider-aws/lib/data-aws-alb"
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { EcsTaskDefinition, } from "@cdktf/provider-aws/lib/ecs-task-definition"
import { EcsService, } from "@cdktf/provider-aws/lib/ecs-service"
import { LbTargetGroup } from "@cdktf/provider-aws/lib/lb-target-group"
import { LbListenerRule } from "@cdktf/provider-aws/lib/lb-listener-rule"
import { LbListener } from "@cdktf/provider-aws/lib/lb-listener"
import { CloudwatchLogGroup } from "@cdktf/provider-aws/lib/cloudwatch-log-group"

type ECSComputeLimit = { cpu: ".25 vcpu", mem: "512" } |
{ cpu: ".25 vcpu", mem: "1 GB" } |
{ cpu: ".25 vcpu", mem: "2 GB" } |
{ cpu: ".5 vcpu", mem: "1 GB" } |
{ cpu: ".5 vcpu", mem: "2 GB" } |
{ cpu: ".5 vcpu", mem: "3 GB" } |
{ cpu: ".5 vcpu", mem: "4 GB" } |
{ cpu: "1 vcpu", mem: "2 GB" } |
{ cpu: "1 vcpu", mem: "3 GB" } |
{ cpu: "1 vcpu", mem: "4 GB" } |
{ cpu: "1 vcpu", mem: "5 GB" } |
{ cpu: "1 vcpu", mem: "6 GB" } |
{ cpu: "1 vcpu", mem: "7 GB" } |
{ cpu: "1 vcpu", mem: "8 GB" } |
{ cpu: "2 vcpu", mem: "4 GB" } |
{ cpu: "2 vcpu", mem: "5 GB" } |
{ cpu: "2 vcpu", mem: "6 GB" } |
{ cpu: "2 vcpu", mem: "7 GB" } |
{ cpu: "2 vcpu", mem: "8 GB" } |
{ cpu: "2 vcpu", mem: "9 GB" } |
{ cpu: "2 vcpu", mem: "10 GB" } |
{ cpu: "2 vcpu", mem: "11 GB" } |
{ cpu: "2 vcpu", mem: "12 GB" } |
{ cpu: "2 vcpu", mem: "13 GB" } |
{ cpu: "2 vcpu", mem: "14 GB" } |
{ cpu: "2 vcpu", mem: "15 GB" } |
{ cpu: "2 vcpu", mem: "16 GB" } |
{ cpu: "1 vcpu", mem: "4 GB" }

enum ServiceType {
    Private,
    Internal,
    Public
}

type Environment = "rdev" | "dev" | "staging" | "prod"

type AWSRegion = "us-east-1" | "us-east-2" | "us-west-1" | "us-west-2"

interface ECSServiceDefinition {
    name: string
    // the number of replicas
    desiredCount: number
    // the port to expose
    port: number
    // the image to use in the container
    image: string
    // deployment size (only valid combinations allowed)
    computeLimits: ECSComputeLimit
    // if the service is on the internet, protected by Okta, or only exposed within the cluster
    serviceType: ServiceType
    healthCheckPath?: string
    environment?: EnvironmentVariables
    command?: string
}

interface EnvironmentVariables { name: string, value: string }[]

interface AWSLogConfig {
    logDriver: "awslogs",
    options: {
        "awslogs-stream-prefix": string,
        "awslogs-group": string,
        "awslogs-region": AWSRegion,
    }
}

interface ContainerDefinition {
    name: string,
    image: string,
    memory: string,
    portMappings: {
        containerPort: number,
        hostPort?: number,
        protocol?: string
    }[]
    logConfiguration?: AWSLogConfig
    essential?: boolean,
    environment?: EnvironmentVariables
    command?: string
}


function makeName(env: Environment, stackName: string, serviceDef: ECSServiceDefinition): string {
    return [env, stackName, serviceDef.name].join("-")
}

class HappyLoadBalancer extends Construct {
    readonly targetGroup: LbTargetGroup
    constructor(
        scope: Construct,
        id: string,
        vpcID: string,
        priority: number,
        hostMatch: string,
        serviceDef: ECSServiceDefinition
    ) {
        super(scope, id)

        this.targetGroup = new LbTargetGroup(scope, "lb_target_group", {
            vpcId: vpcID,
            port: serviceDef.port,
            protocol: "HTTP",
            targetType: "ip",
            deregistrationDelay: "10",
            healthCheck: {
                interval: 15,
                path: "/",
                protocol: "HTTP",
                timeout: 5,
                healthyThreshold: 2,
                unhealthyThreshold: 10,
                matcher: "200-299"
            }
        })

        const lbListener = new LbListener(scope, "lb_listener", {})

        new LbListenerRule(scope, "lb_listener_rule", {
            listenerArn: lbListener.arn,
            priority: priority,
            condition: [{
                hostHeader: { values: [hostMatch] },
            }],
            action: [{
                targetGroupArn: this.targetGroup.arn,
                type: "forward"
            }]
        })
    }
}

class HappyECSTaskDefinition extends Construct {
    readonly taskDef: EcsTaskDefinition
    readonly containerDefs: ContainerDefinition[]

    constructor(
        scope: Construct,
        id: string,
        env: Environment,
        stackName: string,
        region: AWSRegion,
        serviceDef: ECSServiceDefinition
    ) {
        super(scope, id)

        const logGroup = new CloudwatchLogGroup(scope, stackName, {
            retentionInDays: 365,
            name: `/happy/${env}/${stackName}/${serviceDef.name}`
        })

        this.containerDefs = [{
            name: serviceDef.name,
            memory: serviceDef.computeLimits.mem,
            image: serviceDef.image,
            essential: true,
            environment: serviceDef.environment,
            portMappings: [{ containerPort: serviceDef.port }],
            logConfiguration: {
                logDriver: 'awslogs',
                options: {
                    "awslogs-group": logGroup.id,
                    "awslogs-region": region,
                    "awslogs-stream-prefix": env,
                }
            },
            command: serviceDef.command,
        }]

        this.taskDef = new EcsTaskDefinition(scope, stackName, {
            family: makeName(env, stackName, serviceDef),
            memory: serviceDef.computeLimits.mem,
            cpu: serviceDef.computeLimits.cpu,
            networkMode: "awsvpc",
            requiresCompatibilities: ["FARGATE"],
            taskRoleArn: "TODO",
            executionRoleArn: "TODO",
            containerDefinitions: JSON.stringify(this.containerDefs),
        })
    }
}

class HappyECSFargateService extends Construct {
    constructor(
        scope: Construct,
        id: string,
        env: Environment,
        stackName: string,
        serviceDef: ECSServiceDefinition,
        ecsClusterARN: string,
        def: HappyECSTaskDefinition,
        lb: HappyLoadBalancer
    ) {
        super(scope, id)

        new EcsService(scope, "ecs_fargate_service", {
            cluster: ecsClusterARN,
            desiredCount: serviceDef.desiredCount,
            taskDefinition: def.taskDef.arn,
            launchType: "FARGATE",
            name: makeName(env, stackName, serviceDef),
            loadBalancer: def.containerDefs.map(c => ({
                containerName: c.name,
                containerPort: c.portMappings[0].containerPort,
                targetGroupArn: lb.targetGroup.arn
            })),
            networkConfiguration: {
                securityGroups: [],
                subnets: [],
                assignPublicIp: false,
            },
            enableExecuteCommand: true,
            waitForSteadyState: true,
        })
    }
}
