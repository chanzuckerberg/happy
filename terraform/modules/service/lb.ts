
import { Construct } from "constructs";
import { LbListener } from "@cdktf/provider-aws/lib/lb-listener"
import { Lb } from "@cdktf/provider-aws/lib/lb"
import { DataAwsRoute53Zone } from "@cdktf/provider-aws/lib/data-aws-route53-zone";
import { Pet } from "@cdktf/provider-random/lib/pet";

import { AwsAcmCertificate } from '../.gen/modules/aws-acm-certificate';
import { HappyServiceMeta } from "./types"
import { makeName } from "./main";

export interface HappyLoadBalancer {
    readonly lbName: string
    readonly lbZoneID: string
    readonly recordName: string
    readonly listener: LbListener
    readonly baseZone: DataAwsRoute53Zone
}
interface HappyLoadBalancerProps {
    meta: HappyServiceMeta
    baseZoneName: string
    tags?: { [key: string]: string }
}
export class HappyInternalLoadBalancer extends Construct implements HappyLoadBalancer {
    readonly lbName: string
    readonly lbZoneID: string
    readonly recordName: string
    readonly baseZone: DataAwsRoute53Zone
    readonly listener: LbListener
    constructor(scope: Construct, id: string, config: HappyLoadBalancerProps) {
        super(scope, id)

        // This requires the *.internal.* as part of the name so that the nginx/oauth proxy
        // can properly route traffic to this load balancer. This zone will already exist
        // from the happy environment's multi-domain-oauth-proxy.
        this.baseZone = new DataAwsRoute53Zone(scope, "data_record", {
            name: `internal.${config.baseZoneName}`
        })
        this.recordName = `${config.meta.stackName}-${config.meta.serviceDef.name}.${this.baseZone.name}`
        const pet = new Pet(scope, "petlb", { prefix: makeName(config.meta) })
        const lb = new Lb(scope, "lb", {
            name: pet.id,
            internal: true,
            securityGroups: [], // TODO
            subnets: [], //TODO
            idleTimeout: 30,
            tags: config.tags,
        })

        this.lbName = lb.name
        this.lbZoneID = lb.zoneId
        this.listener = new LbListener(scope, "lb_listener", {
            loadBalancerArn: lb.arn,
            port: 80,
            protocol: "HTTP",
            defaultAction: [{
                type: "fixed-response",
                fixedResponse: {
                    contentType: "text/plain",
                    messageBody: "Not Found",
                    statusCode: "404",
                }
            }],
            tags: config.tags
        })
    }
}

export class HappyExternalLoadBalancer extends Construct implements HappyLoadBalancer {
    readonly lbName: string
    readonly lbZoneID: string
    readonly baseZone: DataAwsRoute53Zone
    readonly listener: LbListener
    readonly recordName: string

    constructor(scope: Construct, id: string, config: HappyLoadBalancerProps) {
        super(scope, id)

        const pet = new Pet(scope, "petlb", { prefix: makeName(config.meta) })
        const lb = new Lb(scope, "lb", {
            name: pet.id,
            internal: false,
            securityGroups: [], // TODO
            subnets: [], //TODO
            idleTimeout: 30,
            tags: config.tags,
        })
        this.lbName = lb.name
        this.lbZoneID = lb.zoneId
        this.baseZone = new DataAwsRoute53Zone(scope, "data_record", {
            name: config.baseZoneName
        })
        this.recordName = `${config.meta.serviceDef.name}.${this.baseZone.name}`
        const cert = new AwsAcmCertificate(scope, "lb_cert", {
            awsRoute53ZoneId: "TODO",
            certDomainName: this.recordName,
            tags: config.tags,
        })
        this.listener = new LbListener(scope, "lb_listener", {
            loadBalancerArn: lb.arn,
            port: 443,
            protocol: "HTTPS",
            sslPolicy: "ELBSecurityPolicy-TLS-1-2-Ext-2018-06",
            certificateArn: cert.arnOutput,
            defaultAction: [{
                type: "fixed-response",
                fixedResponse: {
                    contentType: "text/plain",
                    messageBody: "Not Found",
                    statusCode: "404",
                }
            }],
            tags: config.tags
        })
    }
}
