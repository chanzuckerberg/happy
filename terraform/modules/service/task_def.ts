import { CloudwatchLogGroup } from "@cdktf/provider-aws/lib/cloudwatch-log-group";
import { EcsTaskDefinition } from "@cdktf/provider-aws/lib/ecs-task-definition";
import { Construct } from "constructs";
import { makeName } from "./main";
import { ContainerDefinition, HappyServiceMeta } from "./types";

export class HappyECSTaskDefinition extends Construct {
    readonly taskDef: EcsTaskDefinition
    readonly containerDefs: ContainerDefinition[]

    constructor(
        scope: Construct,
        id: string,
        meta: HappyServiceMeta
    ) {
        super(scope, id)

        const logGroup = new CloudwatchLogGroup(scope, meta.stackName, {
            retentionInDays: 365,
            name: `/happy/${meta.env}/${meta.stackName}/${meta.serviceDef.name}`
        })

        this.containerDefs = [{
            name: meta.serviceDef.name,
            memory: meta.serviceDef.computeLimits.mem,
            image: meta.serviceDef.image,
            essential: true,
            environment: meta.serviceDef.environment,
            portMappings: [{ containerPort: meta.serviceDef.port }],
            logConfiguration: {
                logDriver: 'awslogs',
                options: {
                    "awslogs-group": logGroup.id,
                    "awslogs-region": meta.region,
                    "awslogs-stream-prefix": meta.env,
                }
            },
        }]

        this.taskDef = new EcsTaskDefinition(scope, meta.stackName, {
            family: makeName(meta),
            memory: meta.serviceDef.computeLimits.mem,
            cpu: meta.serviceDef.computeLimits.cpu,
            networkMode: "awsvpc",
            requiresCompatibilities: ["FARGATE"],
            taskRoleArn: "TODO",
            executionRoleArn: "TODO",
            containerDefinitions: JSON.stringify(this.containerDefs),
        })
    }
}
