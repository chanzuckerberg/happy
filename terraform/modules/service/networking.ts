import { Construct } from "constructs";
import { LbTargetGroup } from "@cdktf/provider-aws/lib/lb-target-group"
import { LbListenerRule } from "@cdktf/provider-aws/lib/lb-listener-rule"
import { HappyDNS } from "../dns/main";
import { HappyServiceMeta } from "./types"
import { HappyExternalLoadBalancer, HappyInternalLoadBalancer, HappyLoadBalancer } from "./lb";

interface HappyNetworkingProps {
    vpcID: string,
    baseZoneName: string,
    healthCheckPath: string,
    meta: HappyServiceMeta,
    tags?: { [key: string]: string }
}
export class HappyNetworking extends Construct {
    readonly lb: HappyLoadBalancer
    readonly targetGroup: LbTargetGroup
    constructor(
        scope: Construct,
        id: string,
        config: HappyNetworkingProps,
    ) {
        super(scope, id)

        if (config.meta.serviceDef.serviceType === "EXTERNAL") {
            this.lb = new HappyExternalLoadBalancer(scope, "ext_lb", { meta: config.meta, baseZoneName: config.baseZoneName })
        } else {
            this.lb = new HappyInternalLoadBalancer(scope, "int_lb", { meta: config.meta, baseZoneName: config.baseZoneName })
        }

        // the DNS record that will be used to hit this service. It will point to the happy load balancer above
        const dns = new HappyDNS(
            scope,
            "record",
            this.lb.lbName,
            this.lb.lbZoneID,
            this.lb.recordName,
            this.lb.baseZone.name
        )

        // health checking and port routing to the specific container
        this.targetGroup = new LbTargetGroup(scope, "lb_target_group", {
            vpcId: config.vpcID,
            port: config.meta.serviceDef.port,
            protocol: "HTTP",
            targetType: "ip",
            deregistrationDelay: "10",
            healthCheck: {
                interval: 15,
                path: config.healthCheckPath,
                protocol: "HTTP",
                timeout: 5,
                healthyThreshold: 2,
                unhealthyThreshold: 10,
                matcher: "200-299"
            }
        })

        // if the host matches the record that was created, take the request to the container
        new LbListenerRule(scope, "lb_listener_rule", {
            listenerArn: this.lb.listener.arn,
            condition: [{
                hostHeader: { values: [dns.record.name] },
            }],
            action: [{
                targetGroupArn: this.targetGroup.arn,
                type: "forward"
            }]
        })
    }
}
