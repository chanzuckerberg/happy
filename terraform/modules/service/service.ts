import { EcsService } from "@cdktf/provider-aws/lib/ecs-service";
import { Construct } from "constructs";
import { makeName } from "./main";
import { HappyNetworking } from "./networking";
import { HappyECSTaskDefinition } from "./task_def";
import { HappyServiceMeta } from "./types";
import { Pet } from "@cdktf/provider-random/lib/pet";

interface HappyECSFargateServiceProps {
    ecsClusterARN: string,
    meta: HappyServiceMeta,
    taskDef: HappyECSTaskDefinition,
    networking: HappyNetworking,
    tags?: { [key: string]: string };
}
export class HappyECSFargateService extends Construct {
    constructor(
        scope: Construct,
        id: string,
        config: HappyECSFargateServiceProps,
    ) {
        super(scope, id)

        const pet = new Pet(scope, "petecsservice", { prefix: makeName(config.meta) })
        new EcsService(scope, "ecs_fargate_service", {
            cluster: config.ecsClusterARN,
            desiredCount: config.meta.serviceDef.desiredCount,
            taskDefinition: config.taskDef.taskDef.arn,
            launchType: "FARGATE",
            name: pet.id,
            loadBalancer: config.taskDef.containerDefs.map(c => ({
                containerName: c.name,
                containerPort: c.portMappings[0].containerPort,
                targetGroupArn: config.networking.targetGroup.arn
            })),
            networkConfiguration: {
                securityGroups: [],
                subnets: [],
                assignPublicIp: false,
            },
            enableExecuteCommand: true,
            waitForSteadyState: true,
        })
    }
}
