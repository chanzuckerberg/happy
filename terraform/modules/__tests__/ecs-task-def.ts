// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import "cdktf/lib/testing/adapters/jest"; // Load types for expect matchers
import { TerraformStack, Testing } from "cdktf";
import { HappyECSTaskDefinition } from "../service/task_def"
import { HappyServiceMeta, ServiceType, ECSComputeLimit } from "../service/types"
import { CloudwatchLogGroup } from "@cdktf/provider-aws/lib/cloudwatch-log-group";
import { EcsTaskDefinition } from "@cdktf/provider-aws/lib/ecs-task-definition";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";

describe("happy service", () => {
  const meta: HappyServiceMeta = {
    env: "rdev",
    stackName: "myStackName",
    region: "us-west-2",
    serviceDef: {
      name: "myServiceName",
      desiredCount: 1,
      port: 8080,
      image: "dockerImage",
      computeLimits: { cpu: ".25 vcpu", mem: "1 GB" } as ECSComputeLimit,
      serviceType: "private" as ServiceType,
      healthCheckPath: "/healthcheck"
    }
  }

  it("creates a AWS Route53 record", () => {
    const construct = Testing.synthScope((scope) => {
      new HappyECSTaskDefinition(scope, "myTaskDefinition", meta);
    })

    expect(construct).toHaveResourceWithProperties(CloudwatchLogGroup, {
      retention_in_days: 365,
      name: "/happy/rdev/myStackName/myServiceName",
    });

    expect(construct).toHaveResourceWithProperties(EcsTaskDefinition, {
      family: "rdev-myStackName-myServiceName",
      memory: "1 GB",
      cpu: ".25 vcpu",
      network_mode: "awsvpc",
      requires_compatibilities: ["FARGATE"],
      execution_role_arn: "TODO",
      task_role_arn: "TODO",
    });
  });
  describe("can synthesize so", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "testHappyService");
    new HappyECSTaskDefinition(stack, "myTaskDefinition", meta);
    new AwsProvider(stack, "AWS", {
      region: "us-west-2",
    });
    const synth = Testing.synth(stack)
    const fullSynth = Testing.fullSynth(stack)

    it("it looks like it did before", () => {
      expect(synth).toMatchSnapshot();
    });

    it("is valid terraform", () => {
      expect(fullSynth).toBeValidTerraform();
    });

    it("plans without error", () => {
      expect(fullSynth).toPlanSuccessfully()
    });
  })
});
