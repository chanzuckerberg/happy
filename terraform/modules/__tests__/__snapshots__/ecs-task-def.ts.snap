// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy service can synthesize so it looks like it did before 1`] = `
"{
  "provider": {
    "aws": [
      {
        "region": "us-west-2"
      }
    ]
  },
  "resource": {
    "aws_cloudwatch_log_group": {
      "cwg-myStackName": {
        "name": "/happy/rdev/myStackName/myServiceName",
        "retention_in_days": 365
      }
    },
    "aws_ecs_task_definition": {
      "ecstd-myStackName": {
        "container_definitions": "[{\\"name\\":\\"myServiceName\\",\\"memory\\":\\"1 GB\\",\\"image\\":\\"dockerImage\\",\\"essential\\":true,\\"portMappings\\":[{\\"containerPort\\":8080}],\\"logConfiguration\\":{\\"logDriver\\":\\"awslogs\\",\\"options\\":{\\"awslogs-group\\":\\"\${aws_cloudwatch_log_group.cwg-myStackName.id}\\",\\"awslogs-region\\":\\"us-west-2\\",\\"awslogs-stream-prefix\\":\\"rdev\\"}}}]",
        "cpu": ".25 vcpu",
        "execution_role_arn": "TODO",
        "family": "rdev-myStackName-myServiceName",
        "memory": "1 GB",
        "network_mode": "awsvpc",
        "requires_compatibilities": [
          "FARGATE"
        ],
        "task_role_arn": "TODO"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "4.39.0"
      }
    }
  }
}"
`;
